server/src/main/java/org/elasticsearch/cluster/metadata/DataLifecycle.java

```diff
@@ -0,0 +1,130 @@
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Holds the Data Lifecycle Management metadata that are configuring how a data stream is managed.
+ */
+public class DataLifecycle implements SimpleDiffable<DataLifecycle>, ToXContentObject {
+
+    private static final boolean FEATURE_FLAG_ENABLED = "true".equals(System.getProperty("es.dlm_feature_flag_enabled"));
+
+    public static boolean isEnabled() {
+        return FEATURE_FLAG_ENABLED;
+    }
+
+    public static final DataLifecycle EMPTY = new DataLifecycle();
+
+    private static final ParseField DATA_RETENTION_FIELD = new ParseField("data_retention");
+
+    private static final ConstructingObjectParser<DataLifecycle, Void> PARSER = new ConstructingObjectParser<>(
+        "lifecycle",
+        false,
+        (args, unused) -> new DataLifecycle((TimeValue) args[0])
+    );
+
+    static {
+        PARSER.declareField(
+            ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> TimeValue.parseTimeValue(p.textOrNull(), DATA_RETENTION_FIELD.getPreferredName()),
+            DATA_RETENTION_FIELD,
+            ObjectParser.ValueType.STRING_OR_NULL
+        );
+    }
+
+    @Nullable
+    private final TimeValue dataRetention;
+
+    public DataLifecycle() {
+        this.dataRetention = null;
+    }
+
+    public DataLifecycle(@Nullable TimeValue dataRetention) {
+        this.dataRetention = dataRetention;
+    }
+
+    public DataLifecycle(long timeInMills) {
+        this(TimeValue.timeValueMillis(timeInMills));
+    }
+
+    @Nullable
+    public TimeValue getDataRetention() {
+        return dataRetention;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        final DataLifecycle that = (DataLifecycle) o;
+        return Objects.equals(dataRetention, that.dataRetention);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(dataRetention);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalTimeValue(dataRetention);
+    }
+
+    // Visible for testing
+    protected DataLifecycle(StreamInput in) throws IOException {
+        dataRetention = in.readOptionalTimeValue();
+    }
+
+    // In case of a cluster in which the flag is partially enabled, we need to ensure that the communication between the nodes will
+    // always work. That's why when dealing with StreamInput/StreamOutput we always read/write. We reset to null if the flag is disabled.
+    public static DataLifecycle read(StreamInput in) throws IOException {
+        DataLifecycle lifecycle = in.readOptionalWriteable(DataLifecycle::new);
```

Comment:
You are right, this assumes that this is always written as an optional, that's why it works so far, but there is no guarantee, I should rename it to `readOptional`, that would be more correct. 

In case the dlm is disabled, I made the conscious choice to always write and read and reset it to null later. I did this to ensure that nodes will always be able to talk to each other, even in a mixed cluster with some nodes with dlm enabled and some with dlm disabled. If we do not do this, then at the first cluster state communication the node will crash. This is what I witnessed during testing. Your previous suggestion would fix this issue during testing.

So, then the only reason to keep this is for a more graceful handling of a mixed cluster. I like the idea that our cluster will not crash badly if someone enabled the feature flag on only one node for example. But if you find it confusing it might not be worth it. What are your thoughts?

Reply:
After an offline discussion, we decided that we should not follow this approach. We will only write the lifecycle if and only if the feature flag is enabled. We do this because this will allows us in the future to change it or remove it completely before the feature goes live. We believe that facilitating the edge case that someone starts a mixed cluster with a feature flag that should be disabled to begin with is not worth sacrificing the freedom to change this feature.